src/libsync/transmissionchecksumvalidator.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) by Klaas Freitag <freitag@owncloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+#include "config.h"
+#include "filesystem.h"
+#include "transmissionchecksumvalidator.h"
+#include "syncfileitem.h"
+#include "propagatorjobs.h"
+#include "configfile.h"
+
+#include <QtConcurrent>
+
+namespace OCC {
+
+TransmissionChecksumValidator::TransmissionChecksumValidator(const QString& filePath, QObject *parent)
+  :QObject(parent),
+    _filePath(filePath)
+{
+
+}
+
+void TransmissionChecksumValidator::setChecksumType( const QByteArray& type )
+{
+    _checksumType = type;
+}
+
+QString TransmissionChecksumValidator::checksumType()
+{
+    return _checksumType;
+}
+
+
+void TransmissionChecksumValidator::uploadValidation( SyncFileItem *item )
+{
+    QString checksumType = _checksumType;
+    if( checksumType.isEmpty() ) {
+        ConfigFile cfg;
+        checksumType = cfg.transmissionChecksum();
+    }
+
+    if( checksumType.isEmpty() || !item ) {
+        // if there is no checksum defined, continue to upload
+        emit validated();
+    } else {
+        _item = item;
+        // Calculate the checksum in a different thread first.
+        connect( &_watcher, SIGNAL(finished()),
+                 this, SLOT(slotUploadChecksumCalculated()));
+        if( checksumType == checkSumMD5C ) {
+            item->_checksum = checkSumMD5C;
+            item->_checksum += ":";
+            _watcher.setFuture(QtConcurrent::run(FileSystem::calcMd5, _filePath));
+
+        } else if( checksumType == checkSumSHA1C ) {
+            item->_checksum = checkSumSHA1C;
+            item->_checksum += ":";
+            _watcher.setFuture(QtConcurrent::run( FileSystem::calcSha1, _filePath));
+        }
+#ifdef ZLIB_FOUND
+        else if( checksumType == checkSumAdlerC) {
+            item->_checksum = checkSumAdlerC;
+            item->_checksum += ":";
+            _watcher.setFuture(QtConcurrent::run(FileSystem::calcAdler32, _filePath));
+        }
+#endif
+        else {
+            // for an unknown checksum, continue to upload
+            emit validated();
+        }
+    }
+}
+
+void TransmissionChecksumValidator::slotUploadChecksumCalculated( )
+{
+    QByteArray checksum = _watcher.future().result();
+
+    if( !checksum.isEmpty() ) {
+        _item->_checksum.append(checksum);
+    } else {
+        _item->_checksum.clear();
```

Comment:
Should that not be an error?   If the file could not be open to read the checksum or anything?  (because this is also likely case of causing a regression)